File: AuctionrBack/Commands/CommandFactory.java
package AuctionrBack.Commands;

import AuctionrBack.Input.LogEntry;
import AuctionrBack.Storage.*;
import AuctionrBack.Commands.Implementation.*;

public class CommandFactory
{
	private static final String CODE_CREATE = "01";
	private static final String CODE_ADD_CREDIT = "06";
	private static final String CODE_ADVERTISE = "03";
	private static final String CODE_BID = "04";
	private static final String CODE_DELETE = "02";
	private static final String CODE_REFUND = "05";


	private UserStorage userStorage;
	private ItemStorage itemStorage;

	public CommandFactory(UserStorage userStorage, ItemStorage itemStorage)
	{
		this.userStorage = userStorage;
		this.itemStorage = itemStorage;
	}

	public Command Create(LogEntry entry)
	{
		String code = entry.TransactionCode();
		String[] args = entry.Arguments();

		if (code.equals(CODE_CREATE))
		{
			return new CreateCommand(args, userStorage);
		}
		else if (code.equals(CODE_DELETE))
		{
			return new DeleteCommand(args, userStorage, itemStorage);
		}
		else if (code.equals(CODE_ADD_CREDIT))
		{
			return new AddCredit(args, userStorage);
		}
		else if (code.equals(CODE_ADVERTISE))
		{
			return new Advertise(args);
		}
		else if (code.equals(CODE_BID))
		{
			return new Bid(args);
		}
		else if (code.equals(CODE_REFUND))
		{
			return new Refund(args);
		}

		return null;	
	}
}

File: FrontEnd/src/commands/CommandFactory.cpp
#include "CommandFactory.h"

#include "../storage/UserFileStorage.h"
#include "../storage/ItemFileStorage.h"
#include "../storage/DailyLogFile.h"

#include "errors/NoCommand.h"
#include "errors/NotFound.h"
#include "errors/MissingPermissions.h"
#include "errors/NotLoggedIn.h"

#include "implementations/Exit.h"
#include "implementations/CreateCommand.h"
#include "implementations/DeleteCommand.h"
#include "implementations/AddCreditCommand.h"
#include "implementations/Login.h"
#include "implementations/Logout.h"
#include "implementations/Advertise.h"
#include "implementations/Bid.h"
#include "implementations/Refund.h"

CommandFactory::CommandFactory(ApplicationState* application, UserStorage* userStorage, ItemStorage* itemStorage, DailyLog* log)
{
	this->application = application;
	this->options = options;

	this->userStorage = userStorage;
	this->itemStorage = itemStorage;
	this->dailyLog = log;
}

CommandFactory::~CommandFactory()
{
}

Command* CommandFactory::Create(std::string input)
{
	Command* command = NULL;

	if (input == "")
	{
		command = new NoCommand();
	}
	else if (input == "login")
	{
		command = new Login(application, userStorage);
	}
	else if (input == "logout")
	{
		command = new Logout(application, dailyLog);
	}
	else if (input == "create")
	{
		command = new CreateCommand(userStorage);
	}
	else if (input == "delete")
	{
		command = new DeleteCommand(application, userStorage);
	}
	else if (input == "addcredit")
	{
		command = new AddCreditCommand(application, userStorage);
	}
	else if (input == "advertise")
	{
		command = new Advertise(application, userStorage, itemStorage);
	}
	else if (input == "bid")
	{
		command = new Bid(application, userStorage, itemStorage);
	}
	else if (input == "refund")
	{
		command = new Refund(userStorage);
	}
	else if (input == "exit")
	{
		command = new Exit(application);
	}
	else
	{
		command = new NotFound(input);
	}

	return PermissionCheck(command);
}

Command* CommandFactory::PermissionCheck(Command* command)
{
	if (command->IsLoginRequired() && !application->IsLoggedIn())
	{
		delete command;
		return new NotLoggedIn();
	}

	User* user = application->LoggedInUser();

	if (command->IsPermitted(user))
	{
		return command;
	}
	else
	{
		delete command;
		return new MissingPermissions();
	}
}
